This project will aim to analyze the eBay listing for used cars. 
The project is explorative, and aims to gain general insighht into data. 
The project is used as a context to practice the Pandas and NumPy libraries for Python.

# 1. Set up and inspection
First we will set up the project: importing the respective libraries, and inspecting the data.

In [1]:
#importing libraries

import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [2]:
# observing the data
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
# observing summary of columns
print(autos.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

We can see that we have all 20 columns, which are mostly non-null. (most null is for notRepairedDamage).

In [4]:
#observing the top rows
print(autos.head(5))

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere 

The rows seem to have certain data types which could be converted to floats/integers for better analysis (e.g price) and odomoter values. 
The date and time is also read out as a string. We can utilise the datetime libraries in order to operate on these dates more efficiently. 

The row of 'notRepairedDamage' can also be converted into boolean, since it is currently stored as 'yes/no'.


Now looking at the describe statistics.

In [5]:
autos.describe(include='all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-23 18:39:34  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count        44905        50000.000000    47320  50000.000000  47242   
unique           8                 NaN        2           NaN    245   
top      limousine                 NaN  manuell           NaN   golf   
freq         12859                 NaN    36993           NaN   4024   
mean           NaN         2005.073280      NaN    116.355920    NaN   
std            NaN          105.712813      NaN    209.216627    NaN   
min            NaN         1000.000000      NaN      0.000000    NaN   
25%            NaN         1999.000000      NaN     70.000000    NaN   
50%            NaN         2003.000000      NaN    105.000000    NaN   
75%            NaN         2008.000000      NaN    150.000000    NaN   
max            NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  monthOfRegistration fuelType       brand notRepairedDamage  \
count       50000         50000.000000    45518       50000             40171   
unique         13                  NaN        7          40                 2   
top     150,000km                  NaN   benzin  volkswagen              nein   
freq        32424                  NaN    30107       10687             35232   
mean          NaN             5.723360      NaN         NaN               NaN   
std           NaN             3.711984      NaN         NaN               NaN   
min           NaN             0.000000      NaN         NaN               NaN   
25%           NaN             3.000000      NaN         NaN               NaN   
50%           NaN             6.000000      NaN         NaN               NaN   
75%           NaN             9.000000      NaN         NaN               NaN   
max           NaN            12.000000      NaN         NaN               NaN   

                dateCreated  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

To be consistentwith formatting we will use snakecase rather than Camelcase throughout the project. Hence we will rename the headers for the columns.

In [6]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [7]:
#use the following rename dictionary to swap column names
mapping = {
    'dateCrawled':'date_crawled',
    'name':'name',
    'seller':'seller',
    'offerType':'offer_type',
    'price':'price',
    'abtest':'abtest',
    'vehicleType':'vehicle_type',
    'yearOfRegistration':'registration_year',
    'gearbox':'gearbox',
    'powerPS':'powerPS',
    'model':'model',
    'odometer':'odometer',
    'monthOfRegistration':'registration_month',
    'fuelType':'fuel_type',
    'brand':'brand',
    'notRepairedDamage':'unrepaired_damage',
    'dateCreated':'ad_created',
    'nrOfPictures':'nr_of_pictures',
    'postalCode':'postal_code',
    'lastSeen':'last_seen'
}

In [8]:
autos.rename(columns=mapping, inplace=True)
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


Columns succesfully renamed.

# 2. Cleaning the data

We wil now look at the data itself, and perform any preliminary cleaning that ought to be done before analysis.

In [9]:
autos.describe()

registration_year       powerPS  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

Looking at the headline figures, we make a the following oservations:
1. The nr_of_pictures is all 0 , indicating either a data collection error (unlikely that for all listings there are 0 pictures on ebay)
2. Registration year will require investigation, as the minimum registration year is seen as 1000 - unlikely , given that the first car was invented in 1885.
3. There might be missing registration months as well, given that theese can only have values 1-12 , whereas we see a minimum values of 0. It is possible that the nulls are stored as -. 
4. The 'autos.describe' returns the summary statistics of the values stored as floats. Comparing this to 'autos.info' we can see that some of the columns, which we'd expect to be useful to interpret as floats/integeres are stored as strings. These for example, include the 'price' and 'odometer' columns, as these include the currency symbols and distance units


In order to prepare the data for analysis, we will remove the unit descriptors from the individual rows (separating them), and convert the columns into floats. 

We also notice that there are ',' (commas) used in order to seperate the 1000s. We will also remove those to allow for float analysis.

First we convert  the price columns to flot figures

In [10]:
#converting the price column from 'object' to 'string' type
autos['price'] = autos['price'].astype(str)

#applying string operations

autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].astype(int)

#inspecting the data after conversion
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Now we convert the odometeer reading column to float values, removing the distance units, and comma 1000s separator. We will also rename the column header from 'odometer' to 'odometer_km' to preserve the information that the values are stored in kilometers

In [11]:
#converting the odometer column from 'object' to 'string' type
autos['odometer'] = autos['odometer'].astype(str)

#applying string operations

autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].astype(int)

#inspecting the data after conversion
autos['odometer'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer, dtype: float64

In [12]:
# 
autos.rename({'odometer' : 'odometer_km'}, axis = 1,  inplace = True)

print(autos.columns)


Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


We have now converted the numeric columns to store numeric float values, rather than objects.

We will now clean the price and odomoter values , and look at outliers.




## Cleaning the price values


Inspecting the price first, we see that we have '0' values:

In [13]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [14]:
autos['price'].value_counts().sort_index(ascending=True)

0           1421
1            156
2              3
3              1
5              2
8              1
9              1
10             7
11             2
12             3
13             2
14             1
15             2
17             3
18             1
20             4
25             5
29             1
30             7
35             1
40             6
45             4
47             1
49             4
50            49
55             2
59             1
60             9
65             5
66             1
            ... 
151990         1
155000         1
163500         1
163991         1
169000         1
169999         1
175000         1
180000         1
190000         1
194000         1
197000         1
198000         1
220000         1
250000         1
259000         1
265000         1
295000         1
299000         1
345000         1
350000         1
999990         1
999999         2
1234566        1
1300000        1
3890000        1
10000000       1
11111111       2
12345678      

We will inspect the most expensive entries (above 1m).



In [15]:
autos.loc[autos['price'] > 1000000]

date_crawled                                               name  \
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   

       seller offer_type     price   abtest vehicle_type  registration_year  \
2897   privat    Angebot  11111111     test    limousine               1973   
7814   privat    Angebot   1300000  control        coupe               1992   
11137  privat    Angebot  10000000  control        coupe               1960   
22947  privat    Angebot   1234566  control        kombi               1999   
24384  privat    Angebot  11111111     test          NaN               1995   
27371  privat    Angebot  12345678  control          NaN               2017   
39377  privat    Angebot  12345678  control          NaN               2018   
39705  privat    Angebot  99999999  control    limousine               1999   
42221  privat    Angebot  27322222  control    limousine               2014   
47598  privat    Angebot  12345678  control    limousine               2001   
47634  privat    Angebot   3890000     test        coupe               2006   

         gearbox  powerPS     model  odometer_km  registration_month  \
2897     manuell       48    escort        50000                   3   
7814         NaN        0       NaN        50000                  12   
11137    manuell      368       NaN       100000                   1   
22947  automatik      190       NaN       150000                   2   
24384        NaN        0       NaN       150000                   0   
27371        NaN       95     punto       150000                   0   
39377    manuell       95       v40       150000                   6   
39705  automatik      224  s_klasse       150000                   9   
42221    manuell      163        c4        40000                   2   
47598    manuell      101    vectra       150000                   3   
47634        NaN      799       NaN         5000                   7   

      fuel_type           brand unrepaired_damage           ad_created  \
2897     benzin            ford              nein  2016-03-12 00:00:00   
7814        NaN  sonstige_autos              nein  2016-04-04 00:00:00   
11137    benzin  sonstige_autos              nein  2016-03-29 00:00:00   
22947    diesel             bmw               NaN  2016-03-22 00:00:00   
24384       NaN      volkswagen               NaN  2016-03-21 00:00:00   
27371       NaN            fiat               NaN  2016-03-09 00:00:00   
39377       NaN           volvo              nein  2016-03-08 00:00:00   
39705    benzin   mercedes_benz               NaN  2016-03-22 00:00:00   
42221    diesel         citroen               NaN  2016-03-08 00:00:00   
47598    benzin            opel              nein  2016-03-31 00:00:00   
47634       NaN  sonstige_autos              nein  2016-04-04 00:00:00   

       nr_of_pictures  postal_code            last_seen  
2897                0        94469  2016-03-12 22:45:27  
7814                0        60598  2016-04-05 11:34:11  
11137               0        73033  2016-04-06 21:18:11  
22947               0        17454  2016-04-02 03:17:32  
24384               0        18519  2016-03-21 1

At this price point we expect sports/collector cars, however we see that there is a number of cars that we do not expect at that price point (such as a golf TDI, or an old volvo v40). We will remove these rows from the datastet ( these could have been either a bad entry in price, or intentional to gain attention).


In [16]:
#dropping the head values of price
autos.drop([2897], inplace = True)
autos.drop([22947], inplace = True)
autos.drop([24384], inplace = True)
autos.drop([27371], inplace = True)
autos.drop([39377], inplace = True)
autos.drop([39705], inplace = True)
autos.drop([42221], inplace = True)
autos.drop([47598], inplace = True)
autos.drop([11137], inplace = True) #overpriced maserati (1 too many zeroes)



Inspecting further the range of values.
Looking at the range of values we see that there is a number of cars with a price of '< 1000\\$' - we will set the minimum price at 1000\$, since it is unlikely that the cars priced belo w the price point are realised offers. The listing include bidding pages, and so it is possible that sellers had set an initial bidding price very low (e.g at 1\$). We wil exclude these lower price ranges, as this would skew the analysis with listing that currently have low bids, or that have just been listed. 

In [17]:
#setting a minimum price of 1000$

autos = autos[autos['price'] > 1000]

#inspecting final set

print(autos['price'].value_counts().sort_index(ascending=True).head(5))
print(autos['price'].value_counts().sort_index(ascending=True).tail(5))

1039     1
1040     1
1049     6
1050    95
1059     1
Name: price, dtype: int64
350000     1
999990     1
999999     2
1300000    1
3890000    1
Name: price, dtype: int64


In [18]:
autos['price'].describe()

count    3.799200e+04
mean     7.575204e+03
std      2.476903e+04
min      1.039000e+03
25%      2.300000e+03
50%      4.499000e+03
75%      8.999000e+03
max      3.890000e+06
Name: price, dtype: float64

#### Findings:
- We note that the current values can also include prices that are currently being bid (which can skew the price downwards, as the final settlement price is still being reached).
- The mean price is 7575.20\$
- Outliers had been removed


## Cleaning the odometer values
We will now inspect the range of odomoter values to ensure that they are within expected ranges.

In [19]:
autos['odometer_km'].describe()

count     37992.000000
mean     122490.919141
std       40899.765790
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [20]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    22776
125000     4300
100000     1852
90000      1560
80000      1326
70000      1153
60000      1093
50000       982
40000       795
30000       747
20000       688
10000       227
5000        493
Name: odometer_km, dtype: int64

#### Findings:
- There are no significant outliers in the odometer value ranges
- There a proportionately more cares with as mileage increases (which is to be expectedm, as ebay is used to sell also used cars)


## Cleaning the Date values


We now turn our attention towards the columns containing date and time information, to ensure that the values are stored are well-behaved.



The columns containing date information are as follows:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website



In [21]:
autos[['date_crawled','last_seen',
      'ad_created','registration_month',
      'registration_year']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 37992 entries, 0 to 49999
Data columns (total 5 columns):
date_crawled          37992 non-null object
last_seen             37992 non-null object
ad_created            37992 non-null object
registration_month    37992 non-null int64
registration_year     37992 non-null int64
dtypes: int64(2), object(3)
memory usage: 1.7+ MB


Inspecting the columns of the autos dataset, we see that'date crawled' , 'last_seen', and 'ad_created' columns store strings.

Let us take a sample of the columns which store these objects.

In [22]:
autos[['date_crawled','last_seen',
      'ad_created']].head(5)

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

We observe that the first 10 characters in each row represent the date input, in each of the columns. The remainder is used for time storage. We will use the first 10 characters in each row, to perform our analysis.



We will now take a close look at the dsitrivution of values in terms of dates, for the dataframe.

In [23]:
# Removing time values from the dataframe
print('Crawl Dates')
print(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True))

#last_seen dates distribution
print('last_seen Dates')
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True))

#ad_created dates distribution
print('ad_created dates')
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True))

Crawl Dates
2016-03-05    0.025716
2016-03-06    0.014003
2016-03-07    0.035455
2016-03-08    0.032349
2016-03-09    0.032481
2016-03-10    0.033270
2016-03-11    0.032954
2016-03-12    0.037640
2016-03-13    0.016056
2016-03-14    0.036771
2016-03-15    0.033507
2016-03-16    0.029032
2016-03-17    0.030164
2016-03-18    0.012924
2016-03-19    0.035113
2016-03-20    0.038245
2016-03-21    0.037324
2016-03-22    0.032559
2016-03-23    0.032059
2016-03-24    0.028796
2016-03-25    0.030296
2016-03-26    0.033139
2016-03-27    0.031454
2016-03-28    0.035350
2016-03-29    0.034060
2016-03-30    0.032954
2016-03-31    0.031401
2016-04-01    0.034560
2016-04-02    0.036376
2016-04-03    0.039245
2016-04-04    0.036771
2016-04-05    0.013240
2016-04-06    0.003238
2016-04-07    0.001500
Name: date_crawled, dtype: float64
last_seen Dates
2016-03-05    0.001105
2016-03-06    0.003553
2016-03-07    0.004501
2016-03-08    0.006264
2016-03-09    0.008923
2016-03-10    0.009686
2016-03-11    0.0

In [24]:
#Analysing the registration year
print(autos['registration_year'].value_counts(normalize=True, dropna=False).sort_index(ascending=True))

1000    0.000026
1001    0.000026
1927    0.000026
1929    0.000026
1931    0.000026
1934    0.000053
1937    0.000105
1938    0.000026
1939    0.000026
1941    0.000053
1943    0.000026
1948    0.000026
1950    0.000026
1951    0.000053
1952    0.000026
1953    0.000026
1954    0.000053
1955    0.000053
1956    0.000105
1957    0.000053
1958    0.000079
1959    0.000158
1960    0.000421
1961    0.000158
1962    0.000105
1963    0.000211
1964    0.000237
1965    0.000447
1966    0.000553
1967    0.000658
          ...   
1999    0.043351
2000    0.050379
2001    0.052748
2002    0.055459
2003    0.064408
2004    0.068041
2005    0.072910
2006    0.069673
2007    0.059407
2008    0.058012
2009    0.054669
2010    0.041667
2011    0.042562
2012    0.034376
2013    0.020925
2014    0.017083
2015    0.009502
2016    0.016398
2017    0.025795
2018    0.010239
2019    0.000026
2800    0.000026
4100    0.000026
4500    0.000026
5000    0.000053
5911    0.000026
6200    0.000026
8888    0.0000

### Observations:
#### Crawl dates
The crawl has been performed from 5th of march to 7th of April, and is roughly evenly distributed, with the exception of the last crawls in april.|

The data was crawled every day from  5 March 2016 and ending at 7 April 2016.

#### Last Seen dates
The last_seen column denotes when the crawler saw thi ad last online. 
There is a signifact fall off of this in 3 days of posting. This is expected as newly posted vehicles need time to sell. This also might indicate the cars close on offers quite commonly in 3/4 days of posting, and the ad is then closed. The Older the timeframe, the less frequent this occurence is. These rows would also include if a listing was re-posted.

#### Ad Created dates
The older the the ad, the less entries obtain. This is to be expected as older postings are taken off / the cars get sold over time.

#### Registration year
There is a portion of cars where the registration date is out of expected values.
These include '1000' & '1001' and years going past 2016.
We also have entires where registration year is '2100+' ; which might be indicating an unregistered vehicle. 
The cars registered at these dates consitute a small portion of the entire dataset, so we will clean it from the set for analysis. 


The data had been crawled in april 2016, however we seee a fair portion of cars registered in 2017 and 2018 - which is providing false information. 


#### Amending the registration year
As acceptable value for registration year we will keeping the range of values from 1900 to 2016.



In [25]:
autos[autos['registration_year'].between(1900,2016)].describe()

price  registration_year       powerPS    odometer_km  \
count  3.660900e+04       36609.000000  36609.000000   36609.000000   
mean   7.680874e+03        2004.043596    130.149963  122262.831544   
std    2.520566e+04           7.017629    197.578657   40981.690248   
min    1.039000e+03        1927.000000      0.000000    5000.000000   
25%    2.300000e+03        2001.000000     82.000000  100000.000000   
50%    4.500000e+03        2005.000000    120.000000  150000.000000   
75%    9.200000e+03        2008.000000    163.000000  150000.000000   
max    3.890000e+06        2016.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        36609.000000         36609.0  36609.000000  
mean             6.060258             0.0  51883.379497  
std              3.567381             0.0  25669.470411  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  31319.000000  
50%              6.000000             0.0  51065.000000  
75%              9.000000             0.0  72654.000000  
max             12.000000             0.0  99998.000000

In [26]:
# inspecting pre-changes
autos['registration_year'].between(1900,2016).describe()

count     37992
unique        2
top        True
freq      36609
Name: registration_year, dtype: object

In [27]:
# amending with range of dates
autos = autos[autos['registration_year'].between(1900,2016) == True]

In [28]:
#verifying applied changes
autos['registration_year'].between(1900,2016).describe()

count     36609
unique        1
top        True
freq      36609
Name: registration_year, dtype: object

In [29]:
#inspecting the distribution of remainging values
autos['registration_year'].value_counts().sort_index(ascending=True)

1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       1
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       3
1959       6
1960      16
1961       6
1962       4
1963       8
1964       9
1965      17
1966      21
1967      25
1968      26
1969      18
        ... 
1987      51
1988     105
1989     119
1990     181
1991     201
1992     208
1993     208
1994     256
1995     418
1996     501
1997     741
1998    1214
1999    1647
2000    1914
2001    2004
2002    2107
2003    2447
2004    2585
2005    2770
2006    2647
2007    2257
2008    2204
2009    2077
2010    1583
2011    1617
2012    1306
2013     795
2014     649
2015     361
2016     623
Name: registration_year, Length: 77, dtype: int64

### Observations

 - We reduced the number of rows in the dataset from 37992 to 36609 - a 3.64% percent reduction in the dataset. This we will denote as a small redcution, and conclude that we still have a dataset that can be properly analyzed.
 
 - We can see a distribution with a long tail, with many cars registered around 2004/2005, with tails going in both directions. As expected there is a long historical tail. 
 
 - There is a number of old cars from 1944 downards - which are possibly collector car listings. 
 
 - 2016 has a spike of listings, and does not follow the decreasing trend from 2005 onwards - this is to be expected, as this bin would contain new rpdocution cars by dealers. 

# 3. Exploring the data

We will now be looking to explore the dataset, that we have just prepared.



Firstly we will look at the the brand data for the ebay car sales, analysing the brand distribution.

In [30]:
autos['brand'].describe()

count          36609
unique            40
top       volkswagen
freq            7711
Name: brand, dtype: object

We se that we have 40 different brands across all listings.
We will be focusing onbucketing of the top 20 most popular brands of cars. 

In [31]:
autos['brand'].value_counts().head(20)

volkswagen        7711
bmw               4623
mercedes_benz     4115
audi              3599
opel              3235
ford              2129
renault           1342
peugeot           1008
fiat               767
skoda              701
seat               633
smart              608
toyota             538
mazda              520
citroen            505
nissan             499
mini               405
hyundai            391
sonstige_autos     382
volvo              326
Name: brand, dtype: int64

This narrows down our analysis whilst also containing the majoruity of the population.

In [32]:
population_count = autos['brand'].value_counts().head(20).sum() / autos['brand'].value_counts().sum()
print('The retained population is', population_count ,'%')

The retained population is 0.9297440520090688 %


We will store the list of the top20 brands in the list 'selected_brands'

In [33]:
#we store the table filtered by top 20 population
selected_brands = list(autos['brand'].value_counts().head(20).index)


## Exploring the values of listing by brand

We will create a dictionary to analyse the mean values of listings, bucketed by brands.

In [34]:
brands_mean_values = {}

for brand in selected_brands:
    brands_mean_values[brand] = autos.loc[autos['brand'] == brand, 'price'].mean()

In [35]:
#sorting my highest to lowest mean
brands_mean_values
brand_mean_values_sorted = sorted(brands_mean_values.items(), key = lambda x: x[1], reverse = True)

for tuple in brand_mean_values_sorted:
    print(tuple)

('sonstige_autos', 28282.25392670157)
('mini', 10715.237037037037)
('audi', 10405.156988052237)
('mercedes_benz', 9377.267314702309)
('bmw', 9191.208955223881)
('volkswagen', 7000.87718843211)
('skoda', 6903.306704707561)
('nissan', 6515.456913827656)
('hyundai', 6301.148337595908)
('volvo', 6278.003067484663)
('ford', 5912.610145608267)
('seat', 5711.921011058452)
('toyota', 5624.578066914498)
('mazda', 5392.401923076923)
('citroen', 4700.8712871287125)
('opel', 4298.587326120556)
('fiat', 4074.8487614080836)
('peugeot', 4043.121031746032)
('smart', 3826.2006578947367)
('renault', 3677.821907600596)




#### Key findings:
- We see that 'mini', 'audi', 'mercedes_bez', and 'bmw' are the price leaers, and there is a large gap between bmw and the next price bucket of volkswagen (which average at around 7000, as measured aganst 9191 for bmw)
- The range of values for the mean is between 10,715 and 3677 (excluding sonstige_autos).
- 'sonstige_autos' is also an entry that is both relatively popular and high valued. This directly transaltes into 'other cars', and may inbclude a number of small, luxury car brands that exist in the market, hence being high on the mean price list. 




## Exploring the mileage
We will create a dictionary to analyse the mileage of listings, bucketed by brands.

In [36]:
brands_mean_mileage = {} #store brand mean values

for brand in selected_brands:
    brands_mean_mileage[brand] = autos.loc[autos['brand'] == brand, 'odometer_km'].mean()

In [37]:
print(brands_mean_mileage)


{'peugeot': 121676.58730158731, 'volkswagen': 125481.77927635844, 'volvo': 136917.17791411042, 'mini': 88679.01234567902, 'seat': 115663.50710900474, 'ford': 119069.98590887741, 'audi': 127453.4592942484, 'toyota': 113382.8996282528, 'mazda': 119451.92307692308, 'skoda': 109643.36661911555, 'bmw': 131841.87756867835, 'citroen': 113673.26732673268, 'hyundai': 101240.40920716112, 'mercedes_benz': 129924.66585662212, 'nissan': 110220.44088176353, 'renault': 120916.54247391953, 'opel': 123588.87171561051, 'fiat': 107327.24902216428, 'smart': 97409.53947368421, 'sonstige_autos': 88494.76439790576}


## Converting data back into Pandas structures

For more detailed analysis, rather than operating on dictionaries, we will import the data into a pandas structure, in particular we will import it into a series, using the pandas constructor commands

First we create the brand-value series

In [38]:
bmv_series = pd.Series(brands_mean_values)
type(bmv_series)

pandas.core.series.Series

In [39]:
bmv_series

audi              10405.156988
bmw                9191.208955
citroen            4700.871287
fiat               4074.848761
ford               5912.610146
hyundai            6301.148338
mazda              5392.401923
mercedes_benz      9377.267315
mini              10715.237037
nissan             6515.456914
opel               4298.587326
peugeot            4043.121032
renault            3677.821908
seat               5711.921011
skoda              6903.306705
smart              3826.200658
sonstige_autos    28282.253927
toyota             5624.578067
volkswagen         7000.877188
volvo              6278.003067
dtype: float64

Now we create the brand-mileage series

In [40]:
bmm_series = pd.Series(brands_mean_mileage)
type(bmm_series)

pandas.core.series.Series

In [41]:
bmm_series

audi              127453.459294
bmw               131841.877569
citroen           113673.267327
fiat              107327.249022
ford              119069.985909
hyundai           101240.409207
mazda             119451.923077
mercedes_benz     129924.665857
mini               88679.012346
nissan            110220.440882
opel              123588.871716
peugeot           121676.587302
renault           120916.542474
seat              115663.507109
skoda             109643.366619
smart              97409.539474
sonstige_autos     88494.764398
toyota            113382.899628
volkswagen        125481.779276
volvo             136917.177914
dtype: float64

Now we will join the two together, since we can use the brand name as key on which we can join them together.
First setting up the dataframe, from the 'bmv_series'.

In [42]:
df = pd.DataFrame(bmv_series, columns =['mean_price'])

In [43]:
df['mileage'] = bmm_series

In [44]:
df

mean_price        mileage
audi            10405.156988  127453.459294
bmw              9191.208955  131841.877569
citroen          4700.871287  113673.267327
fiat             4074.848761  107327.249022
ford             5912.610146  119069.985909
hyundai          6301.148338  101240.409207
mazda            5392.401923  119451.923077
mercedes_benz    9377.267315  129924.665857
mini            10715.237037   88679.012346
nissan           6515.456914  110220.440882
opel             4298.587326  123588.871716
peugeot          4043.121032  121676.587302
renault          3677.821908  120916.542474
seat             5711.921011  115663.507109
skoda            6903.306705  109643.366619
smart            3826.200658   97409.539474
sonstige_autos  28282.253927   88494.764398
toyota           5624.578067  113382.899628
volkswagen       7000.877188  125481.779276
volvo            6278.003067  136917.177914

In [45]:
df.sort_values('mileage')

mean_price        mileage
sonstige_autos  28282.253927   88494.764398
mini            10715.237037   88679.012346
smart            3826.200658   97409.539474
hyundai          6301.148338  101240.409207
fiat             4074.848761  107327.249022
skoda            6903.306705  109643.366619
nissan           6515.456914  110220.440882
toyota           5624.578067  113382.899628
citroen          4700.871287  113673.267327
seat             5711.921011  115663.507109
ford             5912.610146  119069.985909
mazda            5392.401923  119451.923077
renault          3677.821908  120916.542474
peugeot          4043.121032  121676.587302
opel             4298.587326  123588.871716
volkswagen       7000.877188  125481.779276
audi            10405.156988  127453.459294
mercedes_benz    9377.267315  129924.665857
bmw              9191.208955  131841.877569
volvo            6278.003067  136917.177914

Looking at the correlation:

In [46]:
correlation = df['mean_price'].corr(df['mileage'])
print(correlation)

-0.39582889352056844


As expected we see a negative correlation on aggregate, however due to the bucketing, these values will have factors that will drive the correlation higher or lower. This inidicates that brand, or other factors will be at play when determining the price relationship of the car - which is to be expected. 

# Translating categorical data
In order to make the data more readable for an english-speaking audience, we will translate some of the categorical data into english.



We will do this by the means of mapping dictionaries. The below example translates the german 'ja' , 'nein' into english 'yes' and 'no' by the means of mapping dictionaries.


In [47]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
31     2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   
33     2016-03-15 12:25:26  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...   
34     2016-04-05 07:55:07                            Bmw_530d__navi_tuv03/18   
35     2016-03-29 21:52:56                                        Ford_Mondeo   
36     2016-03-16 23:46:31          Ford_Mondeo_2.0_Turnier_TDCi_DPF_Titanium   
37     2016-03-23 16:47:46                 Seat_Altea_XL_2.0_TDI_CR_DPF_Style   
38     2016-03-21 15:51:10                        Volkswagen_Golf_1.4_Special   
39     2016-03-11 13:58:38                 Mercedes_C_180_Kombi__zu_verkaufen   
40     2016-03-07 14:50:03  VW_Golf__4_Cabrio_2.0_GTI_16V___Leder_MFA_Alus...   
41     2016-03-10 10:46:08  Passat_3b_1.9_TDI_Highline__angemeldet_mit_tuv...   
...                    ...                                                ...   
49961  2016-03-14 19:06:17            Golf_2.0_TDI_Sportline_Sonderproduktion   
49962  2016-03-14 17:57:15  Mitsubishi_Space_Star_1__3_L__Bj_2004_Standhei...   
49963  2016-03-26 19:52:59            Mercedes_Benz_B_200_CDI_Special_Edition   
49964  2016-03-10 16:46:43                                 Audi_2_7_TDI_AVANT   
49965  2016-03-11 10:53:51          Opel_Astra_1.6_Lenkradheizung~Sitzheizung   
49966  2016-04-02 19:49:19  Citroën_C1_1.0_**Euro4**TÜV_OKT_2017**Scheiten...   
49967  2016-03-12 11:50:17                      VW_Passat_2_0_TDI_comfortline   
49968  2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969  2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_F

In [48]:
autos['unrepaired_damage'].value_counts()

nein    29389
ja       2185
Name: unrepaired_damage, dtype: int64

In [49]:
map1 = {
    'ja' : 'yes',
    'nein' : 'no'
}

In [56]:
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(map1)

We repeat the process for other columns.

In [51]:
autos


date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
31     2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   
33     2016-03-15 12:25:26  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...   
34     2016-04-05 07:55:07                            Bmw_530d__navi_tuv03/18   
35     2016-03-29 21:52:56                                        Ford_Mondeo   
36     2016-03-16 23:46:31          Ford_Mondeo_2.0_Turnier_TDCi_DPF_Titanium   
37     2016-03-23 16:47:46                 Seat_Altea_XL_2.0_TDI_CR_DPF_Style   
38     2016-03-21 15:51:10                        Volkswagen_Golf_1.4_Special   
39     2016-03-11 13:58:38                 Mercedes_C_180_Kombi__zu_verkaufen   
40     2016-03-07 14:50:03  VW_Golf__4_Cabrio_2.0_GTI_16V___Leder_MFA_Alus...   
41     2016-03-10 10:46:08  Passat_3b_1.9_TDI_Highline__angemeldet_mit_tuv...   
...                    ...                                                ...   
49961  2016-03-14 19:06:17            Golf_2.0_TDI_Sportline_Sonderproduktion   
49962  2016-03-14 17:57:15  Mitsubishi_Space_Star_1__3_L__Bj_2004_Standhei...   
49963  2016-03-26 19:52:59            Mercedes_Benz_B_200_CDI_Special_Edition   
49964  2016-03-10 16:46:43                                 Audi_2_7_TDI_AVANT   
49965  2016-03-11 10:53:51          Opel_Astra_1.6_Lenkradheizung~Sitzheizung   
49966  2016-04-02 19:49:19  Citroën_C1_1.0_**Euro4**TÜV_OKT_2017**Scheiten...   
49967  2016-03-12 11:50:17                      VW_Passat_2_0_TDI_comfortline   
49968  2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969  2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_F

Inspecting the values with object inputs:


In [52]:
print(autos['seller'].value_counts())
print('\n')
print(autos['offer_type'].value_counts())
print('\n')
print(autos['abtest'].value_counts())
print('\n')
print(autos['vehicle_type'].value_counts())
print('\n')
print(autos['gearbox'].value_counts())
print('\n')
print(autos['fuel_type'].value_counts())

privat    36609
Name: seller, dtype: int64


Angebot    36609
Name: offer_type, dtype: int64


test       18879
control    17730
Name: abtest, dtype: int64


limousine     10388
kombi          7424
kleinwagen     6662
bus            3672
cabrio         2839
coupe          2153
suv            1913
andere          287
Name: vehicle_type, dtype: int64


manuell      26301
automatik     9077
Name: gearbox, dtype: int64


benzin     20912
diesel     13267
lpg          556
cng           58
hybrid        37
elektro       16
andere         7
Name: fuel_type, dtype: int64


We find that there are opportunities to translate from german to english. We will again to this by the means of mappings in pandas. 

In [53]:
seller_map = {
    'privat' : 'private'
}

offer_type_map = {
    'Angebot' : 'offer'
}

#'abtest' column is directly translatable so will be omitted

vehicle_type_map = {
    'limousine' : 'limousine',
    'kombi' : 'kombi',
    'kleinwagen' : 'compact',
    'bus' : 'bus',
    'cabrio' : 'convertible',
    'coupe' : 'coupe',
    'suv' : 'suv',
    'andere' : 'other'
}

gearbox_map = {
    'manuell' : 'manual',
    'automatik' : 'automatic'
}

fuel_type_map = {
    'benzin' : 'gasoline',
    'diesel' : 'diesel',
    'lpg' : 'lpg',
    'cng' : 'cng',
    'hybrid' : 'hybrid',
    'elektro' : 'electric',
    'andere' : 'other' 
}

In [54]:
#applying the mapping

autos['seller'] = autos['seller'].map(seller_map)
autos['offer_type'] = autos['offer_type'].map(offer_type_map)
autos['vehicle_type'] = autos['vehicle_type'].map(vehicle_type_map)
autos['gearbox'] = autos['gearbox'].map(gearbox_map)
autos['fuel_type'] = autos['fuel_type'].map(fuel_type_map)

In [57]:
autos


date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
31     2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   
33     2016-03-15 12:25:26  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...   
34     2016-04-05 07:55:07                            Bmw_530d__navi_tuv03/18   
35     2016-03-29 21:52:56                                        Ford_Mondeo   
36     2016-03-16 23:46:31          Ford_Mondeo_2.0_Turnier_TDCi_DPF_Titanium   
37     2016-03-23 16:47:46                 Seat_Altea_XL_2.0_TDI_CR_DPF_Style   
38     2016-03-21 15:51:10                        Volkswagen_Golf_1.4_Special   
39     2016-03-11 13:58:38                 Mercedes_C_180_Kombi__zu_verkaufen   
40     2016-03-07 14:50:03  VW_Golf__4_Cabrio_2.0_GTI_16V___Leder_MFA_Alus...   
41     2016-03-10 10:46:08  Passat_3b_1.9_TDI_Highline__angemeldet_mit_tuv...   
...                    ...                                                ...   
49961  2016-03-14 19:06:17            Golf_2.0_TDI_Sportline_Sonderproduktion   
49962  2016-03-14 17:57:15  Mitsubishi_Space_Star_1__3_L__Bj_2004_Standhei...   
49963  2016-03-26 19:52:59            Mercedes_Benz_B_200_CDI_Special_Edition   
49964  2016-03-10 16:46:43                                 Audi_2_7_TDI_AVANT   
49965  2016-03-11 10:53:51          Opel_Astra_1.6_Lenkradheizung~Sitzheizung   
49966  2016-04-02 19:49:19  Citroën_C1_1.0_**Euro4**TÜV_OKT_2017**Scheiten...   
49967  2016-03-12 11:50:17                      VW_Passat_2_0_TDI_comfortline   
49968  2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969  2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_F

This concludes our translation of categorical data.

## Identifying categorical data

We will now look twards the name column, to see if there are any patterns in the 'name' columkn, to se whether we can extract an additional column.

In [58]:
print(autos['name'].value_counts())
names = 

BMW_320i                                                             67
BMW_318i                                                             66
Volkswagen_Golf_1.4                                                  56
BMW_316i                                                             54
BMW_116i                                                             53
Opel_Corsa_1.2_16V                                                   52
Volkswagen_Golf                                                      47
Volkswagen_Polo                                                      47
Volkswagen_Polo_1.2                                                  46
BMW_320d_DPF_Touring                                                 37
Volkswagen_Passat_Variant_1.9_TDI                                    35
Volkswagen_Golf_1.6                                                  35
BMW_520i                                                             35
Audi_A4_Avant_2.0_TDI_DPF                                       

In order to make the analysis more efficient, we will perform keyword analysis.
We note that:

The seperator used between words is an underscore, rather than a space.
The input is user generated, and can be any string of characters.

In order to perform the keyword analysis we will seperate the entries into the list, and identify the most common keywords on the lists. We will also convert all to lower case to make the analysis consistent, as we do not need to  differentiate between capitalised words for this particular analysis. 

Then for each list we create a dicionary that counts the unique words.

In order to do this, we use the 'Counter' library, which implements a simple function for counting unqie values for lists.



In [70]:
from collections import Counter
results = Counter()
autos['name'].str.lower().str.split('_').apply(results.update)
print(results)


Counter({'': 12817, 'volkswagen': 4496, 'bmw': 4416, 'tdi': 4243, 'mercedes': 3730, 'audi': 3406, '2.0': 3189, 'benz': 3089, 'opel': 3029, 'golf': 2892, 'dpf': 2667, '1.6': 2108, 'ford': 2043, 'vw': 1857, 'tüv': 1810, '1.4': 1798, '1': 1526, 'cdi': 1470, '1.9': 1409, '16v': 1349, 'mit': 1340, '1.2': 1332, 'neu': 1325, 'klima': 1274, 'automatik': 1260, 'renault': 1237, 'c': 1189, 'cabrio': 1155, 'passat': 1146, 's': 1146, 'top': 1144, '2': 1121, 'navi': 1119, 'edition': 1080, 'sport': 1076, 'a4': 1067, '1.8': 1037, '4': 967, 'astra': 963, 'peugeot': 961, 'polo': 941, 'avant': 914, 'variant': 892, 'tuev': 861, 'e': 855, 'smart': 846, 'touring': 796, '5': 791, 'corsa': 785, 'leder': 775, 'a6': 763, '/': 760, 'a3': 758, 'fiat': 742, 'coupe': 737, 't': 729, 'quattro': 728, 'xenon': 715, 'zustand': 699, 'aut.': 686, 'skoda': 679, 'ahk': 672, 'line': 663, 'kombi': 661, 'seat': 630, '200': 628, '6': 614, 'dsg': 611, 'm': 595, 'diesel': 592, 'tronic': 581, 'focus': 579, 'mini': 572, 'tsi': 565,

We can see that the name will most often state the brand and make of the car. The list also counts popular german used to construct the ads. 

However, we also see words like 'klima', 'neu', and a number of '1.9', '2.5' , etc.

This could indicate that we can potentially create categories for cars with AC, cars that are Brand New, and categorise by Cylinder Volume - as currently these columns do not exist. 

For the time being however, we will takes this out of the scope of the project, and note the possiblity for future exploration.

This can also be taken as a suggestion for the German eBay interfrace for car listings - as there are no particualr columns for this information, people commonly implement the information into the advertisement name, which might indicate that there is demand for such information. 



# Further Analysis

In the final part of the project we will conduct analysis on the following questions:

What are the most common brand/model combinations? <br>
What is the distribution of prices around odometer values? <br>
How do prices differ between damaged and non-damaged cars?
